services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -13,38 +13,89 @@ import scala.concurrent.{ ExecutionContext, Future }
+    optionProjectVersion: Option[PipelineVersion],
+    optionUserVersion: Option[PipelineVersion]
+  ): Either[VersioningException, PipelineVersion] =
+    (optionProjectVersion, optionUserVersion) match {
+      case (Some(projectVersion), Some(userVersion)) =>
+        if (projectVersion >= userVersion)
+          Left(
+            VersioningException(
+              s"Your version $userVersion is out of date. Current version of project: $projectVersion"
+            )
+          )
+        else Right(userVersion)
+      case (Some(projectVersion), None) => Right(projectVersion.increaseRevision)
+      case (None, Some(userVersion))    => Right(userVersion)
+      case (None, None) =>
+        Left(
+          VersioningException(
+            "Can't take decision what version should it be: no version from user and no version in project yet"
+          )
+        )
+    }
+
+  private def handleCreateTag(repositoryId: Repository, version: PipelineVersion, responseBody: String)(
+    implicit ec: ExecutionContext
+  ): AsyncResult[String] =
+    createTag(repositoryId, version).map {
+      case Right(_)        => Right(responseBody)
+      case Left(exception) => Left(exception)
+    }
+
+  override def updateFile(project: Project, projectFile: ProjectFile, userVersion: Option[PipelineVersion])(
     implicit ec: ExecutionContext
   ): AsyncResult[String] = {
     val path = URLEncoder.encode(projectFile.path.toString, "UTF-8")
-    val fileUrl = s"${config.url}projects/${project.repository}/repository/files/$path"
-    val versionValue = version.map(_.name).getOrElse(config.defaultFileVersion)
+    val repositoryId: Repository =
+      project.repository.getOrElse(throw VersioningException(s"No repository for project: $project"))
+    val fileUrl = s"${config.url}projects/${repositoryId.value}/repository/files/$path"
 
+    getLastProjectVersion(project).map(projectVersion => getNewProjectVersion(projectVersion, userVersion)).flatMap {
+      case Left(error) =>
+        Future.successful(Left(error))
+      case Right(newVersion) =>
+        val payload =
+          Json.stringify(
+            Json.toJson(UpdateFileRequest(projectFile.content, newVersion.toString, config.defaultBranch))
+          )
+        httpClient.put(fileUrl, payload = payload, headers = config.token).flatMap {
+          case Response(HttpStatusCodes.OK, body, _) =>
+            handleCreateTag(repositoryId, newVersion, body)
+          case _ =>
+            httpClient.post(fileUrl, payload = payload, headers = config.token).flatMap {
+              case Response(HttpStatusCodes.OK, body, _) =>
+                handleCreateTag(repositoryId, newVersion, body)
+              case Response(_, body, _) => Future.successful(Left(VersioningException(body)))
+            }
+        }
+    }
+  }
+
+  private def createTag(projectId: Repository, version: PipelineVersion)(
+    implicit ec: ExecutionContext
+  ): AsyncResult[String] = {
+    val emptyPayload = "";
```

Comment:
```suggestion
    val emptyPayload = ""
```
Didn't notice it last time.

Reply:
Omg, strange mistake from me